DataStructures.Tests/AATreeTests.cs

```diff
@@ -0,0 +1,221 @@
+            var actual = tree.GetKeysInOrder();
+            Assert.IsTrue(Enumerable.SequenceEqual(expected, actual));
+
+            Assert.DoesNotThrow(() => tree.Validate());
+        }
+
+        [Test]
+        public static void Add_MultipleKeys_FormsCorrectTree()
+        {
+            var tree = new AATree<int>();
+
+            foreach(int elem in new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9 ,10 })
+            {
+                tree.Add(elem);
+                Assert.AreEqual(elem, tree.Count);
+                Assert.IsTrue(tree.Contains(elem));
+            }
+
+            var expected = new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
+            var actual = tree.GetKeysInOrder();
+            Assert.IsTrue(Enumerable.SequenceEqual(expected, actual));
+
+            expected = new int[] { 1, 3, 2, 5, 7, 10, 9, 8, 6, 4 };
+            actual = tree.GetKeysPostOrder();
+            Assert.IsTrue(Enumerable.SequenceEqual(expected, actual));
+
+            Assert.DoesNotThrow(() => tree.Validate());
+        }
+
+        [Test]
+        public static void Add_KeyAlreadyInTree_ThrowsException()
+        {
+            var tree = new AATree<int>();
+            tree.AddRange(new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 });
+
+            Assert.Throws<ArgumentException>(() => tree.Add(1));
+        }
+
+        [Test]
+        public static void AddRange_MultipleKeys_FormsCorrectTree()
+        {
+            var tree = new AATree<int>();
+            tree.AddRange(new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 });
+
+            Assert.AreEqual(10, tree.Count);
+
+            var expected = new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
+            var actual = tree.GetKeysInOrder();
+            Assert.IsTrue(Enumerable.SequenceEqual(expected, actual));
+
+            expected = new int[] { 1, 3, 2, 5, 7, 10, 9, 8, 6, 4 };
+            actual = tree.GetKeysPostOrder();
+            Assert.IsTrue(Enumerable.SequenceEqual(expected, actual));
+
+            Assert.DoesNotThrow(() => tree.Validate());
+        }
+
+        [Test]
+        public static void Remove_MultipleKeys_TreeStillValid()
+        {
+            var tree = new AATree<int>();
+            tree.AddRange(new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 });
+
+            Assert.IsTrue(tree.Remove(8));
+            Assert.IsFalse(tree.Contains(8));
+            Assert.AreEqual(9, tree.Count);
+
+            Assert.IsTrue(tree.Remove(4));
+            Assert.IsFalse(tree.Contains(4));
+            Assert.AreEqual(8, tree.Count);
+
+            Assert.IsTrue(tree.Remove(1));
+            Assert.IsFalse(tree.Contains(1));
+            Assert.AreEqual(7, tree.Count);
+
+            Assert.DoesNotThrow(() => tree.Validate());
+        }
+
+        [Test]
+        public static void Remove_KeyNotInTree_ReturnsFalse()
+        {
+            var tree = new AATree<int>();
+            tree.AddRange(new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 });
```

Comment:
Codacy found an issue: [Remove the array type; it is redundant.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6367315)